src/lib/coda_lib/coda_lib.ml

```diff
@@ -735,6 +738,68 @@ let create (config : Config.t) ~genesis_ledger ~base_proof =
                 in
                 f ~frontier input )
           in
+          (* knot-tying hack so we can pass a get_telemetry function before net created *)
+          let net_ref = ref None in
+          let block_producers =
+            config.initial_block_production_keypairs |> Keypair.Set.to_list
+            |> List.map ~f:(fun {Keypair.public_key; _} ->
+                   Public_key.compress public_key )
+          in
+          let get_telemetry_data _env =
+            let node = config.gossip_net_params.addrs_and_ports.external_ip in
+            match !net_ref with
+            | None ->
+                (* essentially unreachable; without a network, we wouldn't receive this RPC call *)
+                Logger.info config.logger
+                  "Network not instantiated when telemetry data requested"
+                  ~module_:__MODULE__ ~location:__LOC__ ;
+                Deferred.return
+                @@ Error
+                     (Error.of_string
+                        (sprintf
+                           !"Node: %{sexp: Unix.Inet_addr.t}, network not \
+                             instantiated when telemetry data requested"
+                           node))
+            | Some net -> (
+              match Broadcast_pipe.Reader.peek frontier_broadcast_pipe_r with
+              | None ->
+                  Deferred.return
+                  @@ Error
+                       (Error.of_string
+                          (sprintf
+                             !"Node: %{sexp: Unix.Inet_addr.t}, could not get \
+                               transition frontier for telemetry data"
+                             node))
+              | Some frontier ->
+                  let%map peers = Coda_networking.peers net in
+                  let protocol_state_hash =
+                    let tip = Transition_frontier.best_tip frontier in
+                    let state =
+                      Transition_frontier.Breadcrumb.protocol_state tip
+                    in
+                    Coda_state.Protocol_state.hash state
+                  in
+                  let ban_statuses =
+                    Trust_system.Peer_trust.peer_statuses config.trust_system
+                  in
+                  let k_block_hashes =
+                    let k_breadcrumbs =
+                      List.take
+                        (Transition_frontier.best_tip_path frontier)
```

Comment:
The `best_tip_path` function actually returns the path from its root to the current best tip. But the root itself isn't included in the path. So it would always return a list of k-1 blocks. I think you can drop the `List.take` here. This function is weird that it doesn't include the root. You might want to add the hash of the root manual.

Reply:
so the current `protocol_state_hash` should be the first item in the list?

Reply:
(We resolved this offline.)